pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,344 @@
+	return getUsableServiceAccountNamePrefix(serviceAccount.Name) + dockercfgToken
+}
+func getTokenSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return getUsableServiceAccountNamePrefix(serviceAccount.Name) + "-token-"
+}
+
+// createDockercfgSecretIfNeeded makes sure at least one ServiceAccountToken secret exists, and is included in the serviceAccount's Secrets list
+func (e *DockercfgController) createDockercfgSecretIfNeeded(serviceAccount *api.ServiceAccount) error {
+
+	// look for an ImagePullSecret in the form
+	dockercfgSecretName := ""
+	foundDockercfgImagePullSecret := false
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		if strings.HasPrefix(pullSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgImagePullSecret = true
+			dockercfgSecretName = pullSecret.Name
+			break
+		}
+	}
+	foundDockercfgMountableSecret := false
+	for _, mountableSecret := range serviceAccount.Secrets {
+		if strings.HasPrefix(mountableSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
+			foundDockercfgMountableSecret = true
+			dockercfgSecretName = mountableSecret.Name
+			break
+		}
+	}
+
+	switch {
+	// if we already have a docker pull secret, simply return
+	case foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
+		return nil
+
+	case foundDockercfgImagePullSecret && !foundDockercfgMountableSecret, !foundDockercfgImagePullSecret && foundDockercfgMountableSecret:
+		staleDecision, err := e.createDockerPullSecretReference(serviceAccount, dockercfgSecretName)
+		if staleDecision || kapierrors.IsConflict(err) {
+			// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
+			return nil
+		}
+
+		return err
+
+	}
+
+	// if we get here, then we need to create a new dockercfg secret
+	dockercfgSecret, err := e.createDockerPullSecret(serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	staleDecision, err := e.createDockerPullSecretReference(serviceAccount, dockercfgSecret.Name)
+	if staleDecision || kapierrors.IsConflict(err) {
+		// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
+		// we do need to clean up our dockercfgSecret.  token secrets are cleaned up by the controller handling service account dockercfg secret deletes
+		if err := e.client.Secrets(dockercfgSecret.Namespace).Delete(dockercfgSecret.Name); err != nil {
+			glog.Error(err)
+		}
+		return nil
+	}
+
+	return err
+}
+
+// createDockerPullSecretReference updates a service account to reference the dockercfgSecret as a Secret and an ImagePullSecret
+func (e *DockercfgController) createDockerPullSecretReference(staleServiceAccount *api.ServiceAccount, dockercfgSecretName string) ( /*isStale*/ bool, error) {
```

Comment:
simplify to just return an error, and return `kapierrors.NewConflict()` in stale cases?